### DISCUSSION TITLE
How can I use useSWRInfinite with GraphQL?

### DISCUSSION THREAD

**Discussion by ghost** [vercel/swr#601] [General]
Mainly how to rewrite the getKey function?

The query is like

```
const GetPosts = gql`
    query GetPosts($index: Int!) {
        posts(sort: "created_at:desc", index: $index) {
            id
        }
    }
`
```

**Comment by tettoffensive** [comment]
I am working on getting it to work with GraphQL as well. What I'm trying is adding the limit and the cursor to the key that I was using back when I was just using useSWR. It's an array with the first element being the query and the rest being any other dependencies.

In your query I don't see any limit or cursor or anything being passed, so I think you'll need that.

What I'm wondering is how to make the subsequent queries as the user scrolls. Is there some way to trigger a particular page to be loaded using mutate? I can't find any examples. I see something called setSize, but I'm not sure how that works.

**Reply by ghost** [reply]
Thanks. 

I just simplified the query. I am using Strapi as backend api server, which uses the `start` and `limit` to paginate (like `offset` or `skip`). 

It works well when using useSWR. But I don't know how to apply in useSWRInfinite.

**Reply by tettoffensive** [reply]
Yea. there's not much in the way of examples to go on

**Reply by arnaudjnn** [reply]
I'm very interested too for an example using GraphQL!

**Comment by quan-ntekka** [comment]
Probably a bit late to respond, but here is an example for the record in typescript.  Using swr v0.3.5
```typescript
useSWRInfinite(
        (
            pageIndex: number,
            previousPageData: any // use a type here that has nextToken and list of items returned by your graphql
        ) => {
            // reached the end
            if (previousPageData && !previousPageData.nextToken) return null;

            const query = `query Get { 
               getPosts(...) {
                   items { ... fields you want to retrieve ... }
                   nextToken
               }
            }`
            // first page, we don't have `previousPageData`
            if (pageIndex === 0) return [query];

            // add the cursor to the API endpoint
            return [
                query,
                previousPageData.nextToken,
            ];
        },
        async (q, nextToken) => {
            const { data } = fetcher(q, {nextToken}); // update to use the graphql fetcher of your choice
            return {
                nextToken: data?.getPosts.nextToken,
                items: data?.getPosts.items,
            };
        }
    );
```

**Comment by ghost** [comment]
Anyone who stumble upon on this discussion, if you have an `offset` in your query input, and you have a fix `limit` for each request you can use the `index` provided by the `getKey` function to paginate your query.

So something like this:

```typescript
const MAX_NUMBER_OF_ENTRY_BY_QUERY = 25

const GET_POST_LIST = `
  query getPostList($postsInput: PostsInput) {
    posts(postsInput: $postsInput) {
      __typename
      id
      content
    }
  }
`

interface PostList {
  posts: Post[]
}

useSWRInfinite<PostList>((index, previousList: PostList) => {
  if (index !== 0 && previousList.posts.length === 0) {
    return null
  }

  const variables = {
    postsInput: {
      offset: index * MAX_NUMBER_OF_ENTRY_BY_QUERY,
      limit: MAX_NUMBER_OF_ENTRY_BY_QUERY
   }
 }

  return [GET_POST_LIST, variables]
}, fetcher)
```
